qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/JobWorkerTest.java

```diff
@@ -56,81 +61,173 @@ public void shouldActivateJob() {
+          .untilAsserted(() -> assertThat(jobHandlerY.getHandledJobs()).hasSize(1));
+    }
 
     // then
     assertThat(jobHandlerX.getHandledJobs())
         .hasSize(1)
         .extracting(ActivatedJob::getKey)
         .contains(jobX);
-
     assertThat(jobHandlerY.getHandledJobs())
         .hasSize(1)
         .extracting(ActivatedJob::getKey)
         .contains(jobY);
   }
 
+  @ParameterizedTest
+  @MethodSource("provideWorkerConfigurators")
+  void shouldFetchOnlySpecifiedVariables(
+      final BiFunction<String, JobWorkerBuilderStep3, JobWorker> configurator) {
+    // given
+    client.createSingleJob(jobType, b -> {}, "{\"a\":1, \"b\":2, \"c\":3,\"d\":4}");
+
+    // when
+    final var fetchVariables = List.of("a", "b");
+    final var jobHandler = new RecordingJobHandler();
+    final var builder =
+        client
+            .getClient()
+            .newWorker()
+            .jobType(jobType)
+            .handler(jobHandler)
+            .fetchVariables(fetchVariables);
+    try (final var ignored = configurator.apply(jobType, builder)) {
+      Awaitility.await("until all jobs are activated")
+          .untilAsserted(() -> assertThat(jobHandler.getHandledJobs()).hasSize(1));
+    }
+
+    // then
+    final ActivatedJob job = jobHandler.getHandledJobs().get(0);
+    assertThat(job.getVariablesAsMap()).isEqualTo(Map.of("a", 1, "b", 2));
+  }
+
+  @Test
+  void shouldStreamAndActivateJobs() {
+    // given - a job created before any stream was registered, i.e. can only be polled
+    client.createSingleJob(jobType, b -> {});
+
+    // when - create a worker that streams, and create a new job after the stream is registered
+    final var jobHandler = new RecordingJobHandler();
+    final var builder =
+        client.getClient().newWorker().jobType(jobType).handler(jobHandler).enableStreaming();
+    try (final var ignored = builder.open()) {
+      awaitStreamRegistered(jobType);
+      client.createSingleJob(jobType, b -> {});
+
+      // then - expect both jobs to be activated
+      Awaitility.await("until all jobs are activated")
+          .untilAsserted(() -> assertThat(jobHandler.getHandledJobs()).hasSize(2));
+    }
+  }
+
   @Test
-  public void shouldFetchOnlySpecifiedVariables() {
+  void shouldRecreateStreamOnGatewayRestart() {
     // given
-    CLIENT_RULE.createSingleJob(jobType, b -> {}, "{\"a\":1, \"b\":2, \"c\":3,\"d\":4}");
+    final var jobHandler = new RecordingJobHandler();
+    final var builder =
+        client.getClient().newWorker().jobType(jobType).handler(jobHandler).enableStreaming();
 
     // when
-    final List<String> fetchVariables = List.of("a", "b");
+    try (final var ignored = builder.open()) {
+      awaitStreamRegistered(jobType);
+      // have to reuse the same gateway config, otherwise we need to recreate the client to point to
+      // the new gateway, which sort of negates the point of this test
+      final var gatewayConfig = CLUSTER.stopGateway();
```

Comment:
:question: Is there a chance that by the time Gateway restarts with the same config, some other tests grabs the same port number? I guess not as we keep incrementing the port number in SocketUtil.

Reply:
As you said, we never reuse ports, so unlikely.